File: [23.12]Design-Pattern/Java/src/behavioral/state/OnState.java
package behavioral.state;

public class OnState implements PowerState {

    private OnState() {
    }

    private static class SingleInstanceHolder {
        private static final OnState INSTANCE = new OnState();
    }

    public static OnState getInstance() {
        return SingleInstanceHolder.INSTANCE;
    }

    @Override
    public void powerButtonPush(Laptop laptop) {
        System.out.println("노트북 전원 OFF");
        laptop.setPowerState(OffState.getInstance());
    }

    @Override
    public void typeButtonPush() {
        System.out.println("타자 입력");
    }

    @Override
    public String toString() {
        return "전원 상태 ON";
    }
}

File: [23.12]Design-Pattern/Kotlin/src/main/kotlin/behavioral/state/OnState.kt
package behavioral.state

object OnState : State {
    override fun powerButtonPush(laptop: Laptop) {
        println("ON -> OFF")
        laptop.setState(OffState)
    }

    override fun typeButtonPush() {
        println("타자 입력")
    }

    override fun toString(): String {
        return "현재 상태 ON"
    }


}